Speechd bridge (nodejs)

andrewn opened the pull request:
An implementation of #39 using JavaScript. It autospawns a `speech-dispatcher` instance in a similar way to the Python API client and communicates over a UNIX socket. I was concerned that [the SSIP protocol](https://devel.freebsoft.org/doc/speechd/ssip.html) be a bit difficult to work with but it seems good. 

I do prefer this approach as nodejs is better suited to this kind of async stuff. We can also reuse the upcoming WebSocket client and any other infrastructure we create for managing node services.

pixelblend commented on services/speech/src/speechd/connect.js:
```diff
@@ -0,0 +1,35 @@
+  } catch (e) {
+    const isConnectionError = ["ENOENT", "ECONNREFUSED"].includes(e.code);
+    if (isConnectionError) {
+      return null;
```
I'm not sure why this returns `null` for a connection error, when it's then conditionally raised later on. Why not have this function throw the connectionError, which can then be optionally caught and suppressed by the function that calls it?

I think that would make this function clearer, and the callee doesn't have to throw it's own exceptions.

pixelblend commented on services/speech/src/speechd/environment.js:
```diff
@@ -0,0 +1,16 @@
+
+const getExecutablePath = async () =>
+  new Promise((resolve, reject) => {
+    exec(`which speech-dispatcher`, (error, stdout) => {
```
I'd never expect this to throw an error, unless `which` wasn't found in the `PATH`. I think it's more likely to return with an empty `stdout` if the `speech-dispatcher` isn't found.

pixelblend replied:
... I am *dead wrong* about this. `which` will return an exit code of `1` if the binary isn't found in `$PATH`.

pixelblend commented on services/speech/src/speechd/errors.js:
```diff
@@ -0,0 +1,9 @@
+class ConnectionError extends Error {
+  constructor() {
+    super("Cannot connect to speech-dispatcher server");
```
Custom Errors 😍

andrewn commented:
@pixelblend, this is ready for re-review please.

I've addressed your previous comments and added some more functionality:

> I'm not sure why this returns null for a connection error, when it's then conditionally raised later on. Why not have this function throw the connectionError, which can then be optionally caught and suppressed by the function that calls it?

Fixed in 1de6762.

> `which` will return an exit code of 1 if the binary isn't found in $PATH.

Fixed in 339c3a.

I've also added the ability listing supported voice types (8fc24982) and specifying the voice type when sending the `speak` command (613750e4).

pixelblend commented on services/speech/src/handleMessage.js:
```diff
@@ -0,0 +1,19 @@
+module.exports = (speech, broker) => async (topic, payload) => {
+  try {
+    switch (topic) {
+      case "speech.command.speak":
```
Don't forget to alter these to be `/` in the code and docs :)
